[WebInspector] make customizable limit for content resources kept for network page of WebInspector

pgorszkowski-igalia opened the pull request:
#### fb4a8f7902cf4b439590dc24596ff821bb901509
<pre>
[WebInspector] make customizable limit for content resources kept for network page of WebInspector
<a href="https://bugs.webkit.org/show_bug.cgi?id=290160">https://bugs.webkit.org/show_bug.cgi?id=290160</a>

Reviewed by Adrian Perez de Castro, Michael Catanzaro, and Devin Rousso.

Without this change, `maximumResourcesContentSize` is hardcoded (200MB) in
`NetworkResourcesData`. On system with limited memory (embedded devices) it
would be good to have smaller value (50MB).

This change provides the web preference `InspectorMaximumResourcesContentSize`
which is used to set a different value depending on the platform (50 MB for
WPE and 200MB for the rest).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/inspector/NetworkResourcesData.cpp:
(WebCore::NetworkResourcesData::NetworkResourcesData):
* Source/WebCore/inspector/NetworkResourcesData.h:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::InspectorNetworkAgent):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
* Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
(WebCore::PageNetworkAgent::PageNetworkAgent):
* Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp:
(WebCore::WorkerNetworkAgent::WorkerNetworkAgent):

Canonical link: <a href="https://commits.webkit.org/293008@main">https://commits.webkit.org/293008@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0042c39d538c34c2f3749a23f782af31eccc487

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17309 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7526 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102790 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48213 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99729 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17603 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25762 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74402 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31594 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100687 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88313 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54751 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13109 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6223 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47656 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90360 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83121 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6300 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104792 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96307 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24764 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18057 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83457 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25136 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84467 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82890 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20874 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27453 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5134 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18362 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24725 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29894 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/119932 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24547 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33662 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27861 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26121 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/NetworkResourcesData.cpp:
```diff
@@ -44,8 +45,24 @@ WTF_MAKE_TZONE_ALLOCATED_IMPL(NetworkResourcesData::ResourceData);
+            if (value)
+                maximumResourcesContentSize = std::min(maximumResourcesContentSizeMB, std::max(maximumSingleResourceContentSizeMB, *value));
+        }
+    });
```
i don't think this is desired for macOS/iOS/etc. so can we please at least wrap this with a `#if PLATFORM(...)`?

ideally we'd have it be a parameter when creating the `NetworkResourcesData` so that this can be plumbed out to a platform-specific `WebInspectorClient` implementation (see `InspectorPageAgent::overrideSetting` for an example of how something somewhat kinda like this has been done in the past)

pgorszkowski-igalia replied:
> i don't think this is desired for macOS/iOS/etc. so can we please at least wrap this with a `#if PLATFORM(...)`?
> 
> ideally we'd have it be a parameter when creating the `NetworkResourcesData` so that this can be plumbed out to a platform-specific `WebInspectorClient` implementation (see `InspectorPageAgent::overrideSetting` for an example of how something somewhat kinda like this has been done in the past)

@dcrousso : do you think it would be useful setting also in case of macOS/iOS/etc, I see it in embedded world but not sure about other platforms. In that case `#if PLATFORM(GTK) || PLATFORM(WPE)` is enough, right?

burg replied:
I understand the embedded use case. I don't think we need this on iOS/macOS and related platforms.

Pulling runtime settings straight from env vars won't work in ports with a strict sandbox, nor is it easy to test the behavior that way. Therefore, I'd request you implement this as a configurable page override setting or platform-specific value.

pgorszkowski-igalia replied:
@burg : the latest implementation uses web preference `InspectorMaximumResourcesContentSize`

dcrousso commented on Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
```diff
@@ -45,7 +45,7 @@ using namespace Inspector;
 
 PageNetworkAgent::PageNetworkAgent(PageAgentContext& context, InspectorClient* client)
-    : InspectorNetworkAgent(context)
+    : InspectorNetworkAgent(context, client->maximumResourcesContentSize())
```
rather than grabbing from the `InspectorClient` now that we have a preference i think we could just do this
```cpp
: InspectorNetworkAgent(context, context.inspectedPage.settings().inspectorMaximumResourcesContentSize())
```

dcrousso commented on Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp:
```diff
@@ -39,7 +39,7 @@ using namespace Inspector;
 
 WorkerNetworkAgent::WorkerNetworkAgent(WorkerAgentContext& context)
-    : InspectorNetworkAgent(context)
+    : InspectorNetworkAgent(context, 200)
```
ditto
```cpp
: InspectorNetworkAgent(context, context.globalScope.settingsValues().inspectorMaximumResourcesContentSize)
```

dcrousso commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -3884,6 +3884,18 @@ InspectorAttachmentSide:
+InspectorMaximumResourcesContentSize:
+  type: uint32_t
+  status: embedder
+  inspectorOverride: true
```
please remove this as `inspectorOverride` is used to allow Web Inspector to override this setting from the UI (i.e. it's not a category for "this setting is used by Web Inspector") <[https://webkit.org/web-inspector/device-settings/](https://webkit.org/web-inspector/device-settings/)>

dcrousso reviewed (approved):
thanks for iterating on this with me! :)

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -1713,6 +1720,24 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+     * consumption by the inspector. The upper limit of this value is 200MB and lower limit is 50MB.
+     * By default the maximum resources content size is 200MB.
+     *
+     */
```
This comment is missing a "Since" tag.

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -1713,6 +1720,24 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+     *
+     * In case of memory limited environments, this setting can be used to minimize memory
+     * consumption by the inspector. The upper limit of this value is 200MB and lower limit is 50MB.
+     * By default the maximum resources content size is 200MB.
```
"In memory-limited environments, this property can be used to reduce the web inspector's memory usage. The value specified is in megabytes, and its valid range goes from 50 to 200 megabytes."

I think you don't need to specify the default value, as that will be done by the documentation generator.

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -1713,6 +1720,24 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+        g_param_spec_uint(
+            "inspector-maximum-resources-content-size",
+            _("The maximum size of stored resources content by inspector"),
+            _("The maximum size of stored resources content by inspector in mega bytes."),
```
megabytes.

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4437,3 +4462,44 @@ webkit_settings_set_webrtc_udp_ports_range(WebKitSettings* settings, const gchar
+ * webkit_settings_get_inspector_maximum_resources_content_size:
+ * @settings: a #WebKitSettings
+ *
+ * Gets the #WebKitSettings:inspector_maximum_resources_content_size property.
```
I see an inconsistency between how the property is linked here and below in the setter, I am not sure which one is correct, but maybe use the same in both comments.

csaavedra commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -4437,3 +4462,44 @@ webkit_settings_set_webrtc_udp_ports_range(WebKitSettings* settings, const gchar
+ * @settings: a #WebKitSettings
+ * @value: Value to be set
+ *
+ * Set the [property@Settings:inspector_maximum_resources_content_size] property.
```
Here.

csaavedra reviewed (changes requested):
Some informal review comments.

mcatanzaro reviewed (commented):
The new runtime setting is useful, but I don't think this is important enough to be worth the cost of a new WPE/GTK API. I would just lower the limit to 50 MB for WPE and move on.

aperezdc reviewed (approved):
@pgorszkowski-igalia I think the new version that adds the preference without adding new public API is better. This looks to me like a quite obscure setting, and if we add at some point API to the WPE and GTK ports to change values of numeric preferences (through `WebKitFeature`) then that will be enough to change it&mdash;although I expect that the default value will be rarely tweaked.

Patch LGTM if nobody has further concerns.

pgorszkowski-igalia committed:
[WebInspector] make customizable limit for content resources kept for network page of WebInspector
https://bugs.webkit.org/show_bug.cgi?id=290160

Reviewed by Adrian Perez de Castro, Michael Catanzaro, and Devin Rousso.

Without this change, `maximumResourcesContentSize` is hardcoded (200MB) in
`NetworkResourcesData`. On system with limited memory (embedded devices) it
would be good to have smaller value (50MB).

This change provides the web preference `InspectorMaximumResourcesContentSize`
which is used to set a different value depending on the platform (50 MB for
WPE and 200MB for the rest).

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/inspector/NetworkResourcesData.cpp:
(WebCore::NetworkResourcesData::NetworkResourcesData):
* Source/WebCore/inspector/NetworkResourcesData.h:
* Source/WebCore/inspector/agents/InspectorNetworkAgent.cpp:
(WebCore::InspectorNetworkAgent::InspectorNetworkAgent):
* Source/WebCore/inspector/agents/InspectorNetworkAgent.h:
* Source/WebCore/inspector/agents/page/PageNetworkAgent.cpp:
(WebCore::PageNetworkAgent::PageNetworkAgent):
* Source/WebCore/inspector/agents/worker/WorkerNetworkAgent.cpp:
(WebCore::WorkerNetworkAgent::WorkerNetworkAgent):

Canonical link: https://commits.webkit.org/293008@main
